File: 1218-longest-arithmetic-subsequence-of-given-difference/1218-longest-arithmetic-subsequence-of-given-difference.java
class Solution {
    public int longestSubsequence(int[] arr, int diff) {
        int n = arr.length;
        
        Map<Integer,Integer> map = new HashMap<>();
        int maxi = 1;
        
        for(int i = 0;i<n;i++){
            int num = arr[i];
            if(map.containsKey(num-diff)){
                map.put(num,map.get(num-diff)+1);
            }
            else{
                map.put(num,1);
            }
            maxi = Math.max(maxi,map.get(num));
        }
        return maxi;
    }
}

File: 1218-longest-arithmetic-subsequence-of-given-difference/1218-longest-arithmetic-subsequence-of-given-difference.cpp
class Solution {
public:
    int longestSubsequence(vector<int>& arr, int diff) {
        unordered_map <int, int > map;
        int maxi=1;
       for(int i =0;i< arr.size();i++){
          int n = arr[i];
           if(map.count(n-diff)>0){
               map[n] = map[n-diff] + 1;
           }else{
               map[n] = 1;

           }
           maxi = max(map[n], maxi);
       }
        return maxi;
       
    
    }
};